File: apps/beny-bot/commands/availability.ts
import { GetAvailabilitiesResponseSchema } from "@planner/schemas/discord";
import axios from "axios";
import {
	type ChatInputCommandInteraction,
	LabelBuilder,
	ModalBuilder,
	type ModalSubmitInteraction,
	StringSelectMenuBuilder,
	StringSelectMenuOptionBuilder,
	TextInputBuilder,
	TextInputStyle,
} from "discord.js";
import { config } from "../lib/config";
import { headers } from "../lib/constants";
import { extractErrorDetails } from "../lib/errorHandler";
import logger from "../lib/logger";
import {
	formatTime,
	getDayName,
	getDayNumber,
	mapStringInputToTime,
} from "../lib/utils";
import { CommandsEnum } from "./types";

async function actionRemove(interaction: ChatInputCommandInteraction) {
	logger.info(
		{ operation: "beny-bot.availability-remove" },
		`Removing an availability timeslot for ${interaction.user.globalName}`,
	);
	const userId = interaction.user.id;
	const dayOfWeek = interaction.options.get("day")?.value;
	const startTime = interaction.options.get("time")?.value;

	if (!(dayOfWeek && startTime)) {
		await interaction.reply({
			content: "I could not parse the day and time given.",
			flags: ["Ephemeral"],
		});
		return;
	}
	const normalizedDay = getDayNumber(dayOfWeek.toString().toLowerCase().trim());
	const normalizedtime = mapStringInputToTime(startTime.toString());
	if (normalizedDay === -1) {
		await interaction.reply({
			content: "I could not read the day you gave me.",
			flags: ["Ephemeral"],
		});
		return;
	}
	if (normalizedtime === null) {
		await interaction.reply({
			content: "I could not read the time you gave me.",
			flags: ["Ephemeral"],
		});
		return;
	}

	try {
		await axios.post(
			`${config.APP_URL}/api/discord/removeAvailability`,
			{
				dayOfWeek: normalizedDay,
				startTime: normalizedtime,
				userExternalId: userId,
			},
			{ headers },
		);

		await interaction.reply({
			content: "Successfully removed availability timeslot.",
			flags: ["Ephemeral"],
		});
	} catch (error) {
		extractErrorDetails({ error, operation: "beny-bot.availability-remove" });
		await interaction.reply({
			content:
				"Something went wrong when trying to remove your availability timeslot. Please try again later or reach out for help.",
			flags: ["Ephemeral"],
		});
	}
}

async function actionClear(interaction: ChatInputCommandInteraction) {
	logger.info(
		{ operation: "beny-bot.availability-clear" },
		`Clearing availability timeslots for ${interaction.user.globalName}`,
	);

	const userId = interaction.user.id;
	try {
		await axios.post(
			`${config.APP_URL}/api/discord/clearAvailability`,
			{
				userExternalId: userId,
			},
			{ headers },
		);

		await interaction.reply({
			content: "Successfully removed all of your availability timeslots.",
			flags: ["Ephemeral"],
		});
	} catch (error) {
		extractErrorDetails({ error, operation: "beny-bot.availability-clear" });
		await interaction.reply({
			content:
				"Something went wrong when trying to remove all of your availability timeslots. Please try again later or reach out for help.",
			flags: ["Ephemeral"],
		});
	}
}

async function actionView(interaction: ChatInputCommandInteraction) {
	logger.info(
		{ operation: "beny-bot.availability-view" },
		`Viewing availabilities for discord user: ${interaction.user.globalName}`,
	);
	const userId = interaction.user.id;
	try {
		const res = await axios.get(
			`${config.APP_URL}/api/discord/getAvailabilities`,
			{
				headers,
				params: { userExternalId: userId },
			},
		);

		const { userAvailabilities } = GetAvailabilitiesResponseSchema.parse(
			res.data,
		);

		if (!userAvailabilities || userAvailabilities.length === 0) {
			await interaction.reply({
				content: "❌ You haven't set any availability yet.",
				ephemeral: true,
			});
			return;
		}

		// sort by day, then time
		const sortedAvailabilities = userAvailabilities.sort((a, b) => {
			if (a.dayOfWeek !== b.dayOfWeek) {
				return a.dayOfWeek - b.dayOfWeek;
			}
			return a.startTime.localeCompare(b.startTime);
		});

		// Format availabilities
		const formattedAvailabilities = sortedAvailabilities
			.map((slot) => {
				const day = getDayName(slot.dayOfWeek);
				const startTime = formatTime(slot.startTime);
				const endTime = formatTime(slot.endTime);
				return `• **${day}**: ${startTime} - ${endTime}`;
			})
			.join("\n");

		await interaction.reply({
			content: `📅 **Your Availability**\n\n${formattedAvailabilities}`,
			flags: ["Ephemeral"],
		});
	} catch (error) {
		extractErrorDetails({ error, operation: "beny-bot.availability-view" });
		await interaction.reply({
			content:
				"Failed to check for your availabilities. Please try again later.",
			flags: ["Ephemeral"],
		});
	}
}

async function actionSet(interaction: ChatInputCommandInteraction) {
	logger.info(
		{ operation: "beny-bot.available-set" },
		"Setting an availability.",
	);
	const setAvailableModal = new ModalBuilder()
		.setCustomId("available:set")
		.setTitle("Set a time you are available");

	const dayOfWeekSelect = new StringSelectMenuBuilder()
		.setCustomId("availabilityDay")
		.setRequired(true)
		.addOptions(
			// matching values to day of week enum
			new StringSelectMenuOptionBuilder()
				.setLabel("Sunday")
				.setValue("0"),
			new StringSelectMenuOptionBuilder().setLabel("Monday").setValue("1"),
			new StringSelectMenuOptionBuilder().setLabel("Tuesday").setValue("2"),
			new StringSelectMenuOptionBuilder().setLabel("Wednesday").setValue("3"),
			new StringSelectMenuOptionBuilder().setLabel("Thursday").setValue("4"),
			new StringSelectMenuOptionBuilder().setLabel("Friday").setValue("5"),
			new StringSelectMenuOptionBuilder().setLabel("Saturday").setValue("6"),
		);

	const startTimeInput = new TextInputBuilder()
		.setCustomId("availabilityStartTime")
		.setStyle(TextInputStyle.Short)
		.setPlaceholder("Examples: 7:00 PM, 19:00, 7pm")
		.setRequired(true)
		.setMaxLength(10);
	const endTimeInput = new TextInputBuilder()
		.setCustomId("availabilityEndTime")
		.setStyle(TextInputStyle.Short)
		.setPlaceholder("Examples: 7:00 PM, 19:00, 7pm")
		.setMaxLength(10);

	const frequencySelect = new StringSelectMenuBuilder()
		.setCustomId("availabilityFrequency")
		.addOptions(
			new StringSelectMenuOptionBuilder().setLabel("Every Week").setValue("1"),
			new StringSelectMenuOptionBuilder()
				.setLabel("Every Other Week")
				.setValue("2"),
		);

	const dayOfWeekLabel = new LabelBuilder()
		.setLabel("Day of Week")
		.setStringSelectMenuComponent(dayOfWeekSelect);
	const hourLabel = new LabelBuilder()
		.setLabel("Start Time")
		.setTextInputComponent(startTimeInput);
	const minuteLabel = new LabelBuilder()
		.setLabel("End Time")
		.setTextInputComponent(endTimeInput);
	const frequencyLabel = new LabelBuilder()
		.setLabel("Frequency")
		.setStringSelectMenuComponent(frequencySelect);

	setAvailableModal.addLabelComponents(
		dayOfWeekLabel,
		hourLabel,
		minuteLabel,
		frequencyLabel,
	);

	await interaction.showModal(setAvailableModal);
}

async function modalSetOnSubmit(interaction: ModalSubmitInteraction) {
	logger.info(
		{ operation: "beny-bot.availability-set" },
		"Setting timeslot for user's availability",
	);
	const serverId = interaction.guildId;
	const userId = interaction.user.id;
	if (!serverId) {
		await interaction.reply({
			content:
				"This command needs to be used inside of a discord server to work.",
			flags: ["Ephemeral"],
		});
		return;
	}
	const day = Number(
		interaction.fields.getStringSelectValues("availabilityDay")[0],
	);
	const frequency = Number(
		interaction.fields.getStringSelectValues("availabilityFrequency")?.[0],
	);
	const startTime = interaction.fields.getTextInputValue(
		"availabilityStartTime",
	);
	const parsedStart = mapStringInputToTime(startTime);
	const endTime = interaction.fields.getTextInputValue("availabilityEndTime");
	const parsedEnd = endTime === "" ? "23:59:59" : mapStringInputToTime(endTime);

	if (parsedStart === null) {
		await interaction.reply({
			content: "I could not convert the start time into a valid time",
			flags: ["Ephemeral"],
		});
		return;
	}
	if (parsedEnd === null) {
		await interaction.reply({
			content: "I could not convert the end time into a valid time",
			flags: ["Ephemeral"],
		});
		return;
	}
	try {
		await axios.post(
			`${config.APP_URL}/api/discord/setAvailability`,
			{
				externalId: userId,
				serverId,
				time: {
					dayOfWeek: day,
					endTime: parsedEnd,
					frequency,
					startTime: parsedStart,
				},
			},
			{ headers },
		);
		await interaction.reply({
			content:
				"Availability set successfully. Use /availability view to see your currently set availability timeslots.",
			flags: ["Ephemeral"],
		});
	} catch (error) {
		const details = extractErrorDetails({
			error,
			operation: "beny-bot.availability-set",
		});

		switch (details.status) {
			case 409: {
				await interaction.reply({
					content:
						"Timeslot overlapping with an already existing one. Use /availability view to see your already set availabilities.",
					flags: ["Ephemeral"],
				});
				break;
			}
			default: {
				await interaction.reply({
					content: "Failed to set availability. Please try again later",
					flags: ["Ephemeral"],
				});
			}
		}
	}
}

export const availabilitySubcommands = [
	{
		action: actionSet,
		description: "Set your recurring availability",
		modal: { id: "available:set", onSubmit: modalSetOnSubmit },
		name: "set",
	},
	{
		action: actionView,
		description: "View your current availability",
		name: "view",
	},
	{
		action: actionRemove,
		description: "Remove a specific availability rule",
		name: "remove",
		options: [
			{
				description: "The Day of the timeslot you'd like removed.",
				isRequired: true,
				name: "day",
			},
			{
				description: "The Time of the timeslot you'd like removed.",
				isRequired: true,
				name: "time",
			},
		],
	},
	{
		action: actionClear,
		description: "Clear all your availability rules",
		name: "clear",
	},
] as const;

export const availabilityCommand = {
	command: CommandsEnum.AVAILABLE,
	description: "Manage your D&D availability",
	subcommands: availabilitySubcommands,
};

File: apps/api/internal/bot/commands/availability.go
package commands

import (
	"fmt"
	"log/slog"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/BBruington/party-planner/api/internal/api"
	"github.com/BBruington/party-planner/api/internal/lib"
	"github.com/bwmarrin/discordgo"
)

type availabilitySlot struct {
	DayOfWeek int    `json:"dayOfWeek"`
	StartTime string `json:"startTime"`
	EndTime   string `json:"endTime"`
}

type getAvailabilitiesResponse struct {
	UserAvailabilities []availabilitySlot `json:"userAvailabilities"`
}

func availabilitySetAction(s *discordgo.Session, i *discordgo.InteractionCreate, client *api.Client) error {
	slog.Info("Setting an availability", "operation", "beny-bot.available-set")

	modal := &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: &discordgo.InteractionResponseData{
			CustomID: "available:set",
			Title:    "Set a time you are available",
			Components: []discordgo.MessageComponent{
				discordgo.ActionsRow{
					Components: []discordgo.MessageComponent{
						&discordgo.TextInput{
							CustomID:    "availabilityDay",
							Label:       "Day of Week",
							Style:       discordgo.TextInputShort,
							Placeholder: "Sunday, Monday, Tuesday, ..., Saturday",
							Required:    true,
							MaxLength:   9,
						},
					},
				},
				discordgo.ActionsRow{
					Components: []discordgo.MessageComponent{
						&discordgo.TextInput{
							CustomID:    "availabilityStartTime",
							Label:       "Start Time",
							Style:       discordgo.TextInputShort,
							Placeholder: "Examples: 7:00 PM, 19:00, 7pm",
							Required:    true,
							MaxLength:   10,
						},
					},
				},
				discordgo.ActionsRow{
					Components: []discordgo.MessageComponent{
						&discordgo.TextInput{
							CustomID:    "availabilityEndTime",
							Label:       "End Time (leave blank for all day)",
							Style:       discordgo.TextInputShort,
							Placeholder: "Examples: 7:00 PM, 19:00, 7pm",
							Required:    false,
							MaxLength:   10,
						},
					},
				},
				discordgo.ActionsRow{
					Components: []discordgo.MessageComponent{
						&discordgo.TextInput{
							CustomID:    "availabilityFrequency",
							Label:       "Frequency",
							Style:       discordgo.TextInputShort,
							Placeholder: "1 = every week, 2 = every other week",
							Required:    true,
							MaxLength:   1,
						},
					},
				},
			},
		},
	}

	return s.InteractionRespond(i.Interaction, modal)
}

func availabilitySetModalOnSubmit(s *discordgo.Session, i *discordgo.InteractionCreate, client *api.Client) error {
	slog.Info("Setting timeslot for user's availability", "operation", "beny-bot.availability-set")

	if i.GuildID == "" {
		return replyEphemeral(s, i, "This command needs to be used inside of a discord server to work.")
	}

	data := i.ModalSubmitData()
	userID := i.Member.User.ID
	serverId := i.GuildID

	dayStr := getModalTextInput(data, "availabilityDay")
	frequencyStr := getModalTextInput(data, "availabilityFrequency")
	startTimeStr := getModalTextInput(data, "availabilityStartTime")
	endTimeStr := getModalTextInput(data, "availabilityEndTime")

	day := lib.GetDayNumber(dayStr)
	if day == -1 {
		return replyEphemeral(s, i, "I could not read the day you entered. Please use a day name like 'Monday'.")
	}

	frequency, err := strconv.Atoi(strings.TrimSpace(frequencyStr))
	if err != nil || (frequency != 1 && frequency != 2) {
		return replyEphemeral(s, i, "Frequency must be 1 (every week) or 2 (every other week).")
	}

	parsedStart, ok := lib.MapStringInputToTime(startTimeStr)
	if !ok {
		return replyEphemeral(s, i, "I could not convert the start time into a valid time")
	}

	var parsedEnd string
	if endTimeStr == "" {
		parsedEnd = "23:59:59"
	} else {
		parsedEnd, ok = lib.MapStringInputToTime(endTimeStr)
		if !ok {
			return replyEphemeral(s, i, "I could not convert the end time into a valid time")
		}
	}

	startTime, err := time.Parse("15:04:05", parsedStart)
	if err != nil {
		return replyEphemeral(s, i, "I could not convert the start time into a valid time")
	}

	endTime, err := time.Parse("15:04:05", parsedEnd)
	if err != nil {
		return replyEphemeral(s, i, "I could not convert the end time into a valid time")
	}

	if !endTime.After(startTime) {
		return replyEphemeral(s, i, "End time must be after start time")
	}

	body := map[string]any{
		"externalId": userID,
		"serverId":   serverId,
		"time": map[string]any{
			"dayOfWeek": day,
			"endTime":   parsedEnd,
			"frequency": frequency,
			"startTime": parsedStart,
		},
	}

	err = client.Post("/api/discord/availability", body, nil)
	if err != nil {
		slog.Error("Failed to set availability", "operation", "beny-bot.availability-set", "error", err)
		if isStatusCode(err, 409) {
			return replyEphemeral(s, i, "Timeslot overlapping with an already existing one. Use /availability view to see your already set availabilities.")
		}
		return replyEphemeral(s, i, "Failed to set availability. Please try again later")
	}

	return replyEphemeral(s, i, "Availability set successfully. Use /availability view to see your currently set availability timeslots.")
}

func availabilityViewAction(s *discordgo.Session, i *discordgo.InteractionCreate, client *api.Client) error {
	if i.GuildID == "" || i.Member == nil || i.Member.User == nil {
		return replyEphemeral(s, i, "This command needs to be used inside of a discord server to work.")
	}

	slog.Info("Viewing availabilities", "operation", "beny-bot.availability-view", "user", i.Member.User.GlobalName)

	userID := i.Member.User.ID
	var result getAvailabilitiesResponse
	err := client.Get("/api/discord/availability", map[string]string{
		"userExternalId": userID,
	}, &result)
	if err != nil {
		slog.Error("Failed to get availabilities", "operation", "beny-bot.availability-view", "error", err)
		return replyEphemeral(s, i, "Failed to check for your availabilities. Please try again later.")
	}

	if len(result.UserAvailabilities) == 0 {
		return replyEphemeral(s, i, "❌ You haven't set any availability yet.")
	}

	// Sort by day then time
	sorted := result.UserAvailabilities
	sort.Slice(sorted, func(a, b int) bool {
		if sorted[a].DayOfWeek != sorted[b].DayOfWeek {
			return sorted[a].DayOfWeek < sorted[b].DayOfWeek
		}
		return strings.Compare(sorted[a].StartTime, sorted[b].StartTime) < 0
	})

	lines := make([]string, 0, len(sorted))
	for _, slot := range sorted {
		day := lib.GetDayName(slot.DayOfWeek)
		start := lib.FormatTime(slot.StartTime)
		end := lib.FormatTime(slot.EndTime)
		lines = append(lines, fmt.Sprintf("• **%s**: %s - %s", day, start, end))
	}

	content := "📅 **Your Availability**\n\n" + strings.Join(lines, "\n")
	return replyEphemeral(s, i, content)
}

func availabilityRemoveAction(s *discordgo.Session, i *discordgo.InteractionCreate, client *api.Client) error {
	if i.GuildID == "" || i.Member == nil || i.Member.User == nil {
		return replyEphemeral(s, i, "This command needs to be used inside of a discord server to work.")
	}

	slog.Info("Removing an availability timeslot", "operation", "beny-bot.availability-remove", "user", i.Member.User.GlobalName)

	userID := i.Member.User.ID
	data := i.ApplicationCommandData()
	var dayVal, timeVal string
	for _, opt := range data.Options[0].Options {
		switch opt.Name {
		case "day":
			dayVal = opt.StringValue()
		case "time":
			timeVal = opt.StringValue()
		}
	}

	if dayVal == "" || timeVal == "" {
		return replyEphemeral(s, i, "I could not parse the day and time given.")
	}

	normalizedDay := lib.GetDayNumber(dayVal)
	if normalizedDay == -1 {
		return replyEphemeral(s, i, "I could not read the day you gave me.")
	}

	normalizedTime, ok := lib.MapStringInputToTime(timeVal)
	if !ok {
		return replyEphemeral(s, i, "I could not read the time you gave me.")
	}

	body := map[string]any{
		"dayOfWeek":      normalizedDay,
		"startTime":      normalizedTime,
		"userExternalId": userID,
	}

	err := client.Delete("/api/discord/availability/single", body)
	if err != nil {
		slog.Error("Failed to remove availability", "operation", "beny-bot.availability-remove", "error", err)
		return replyEphemeral(s, i, "Something went wrong when trying to remove your availability timeslot. Please try again later or reach out for help.")
	}

	return replyEphemeral(s, i, "Successfully removed availability timeslot.")
}

func availabilityClearAction(s *discordgo.Session, i *discordgo.InteractionCreate, client *api.Client) error {
	if i.GuildID == "" || i.Member == nil || i.Member.User == nil {
		return replyEphemeral(s, i, "This command needs to be used inside of a discord server to work.")
	}

	slog.Info("Clearing availability timeslots", "operation", "beny-bot.availability-clear", "user", i.Member.User.GlobalName)

	userID := i.Member.User.ID
	body := map[string]any{
		"userExternalId": userID,
	}

	err := client.Delete("/api/discord/availability", body)
	if err != nil {
		slog.Error("Failed to clear availability", "operation", "beny-bot.availability-clear", "error", err)
		return replyEphemeral(s, i, "Something went wrong when trying to remove all of your availability timeslots. Please try again later or reach out for help.")
	}

	return replyEphemeral(s, i, "Successfully removed all of your availability timeslots.")
}

var AvailabilityCommand = Command{
	Name:        "available",
	Description: "Manage your D&D availability",
	Subcommands: []Subcommand{
		{
			Name:        "set",
			Description: "Set your recurring availability",
			Action:      availabilitySetAction,
			Modal: &Modal{
				ID:       "available:set",
				OnSubmit: availabilitySetModalOnSubmit,
			},
		},
		{
			Name:        "view",
			Description: "View your current availability",
			Action:      availabilityViewAction,
		},
		{
			Name:        "remove",
			Description: "Remove a specific availability rule",
			Action:      availabilityRemoveAction,
			Options: []Option{
				{Name: "day", Description: "The Day of the timeslot you'd like removed.", IsRequired: true},
				{Name: "time", Description: "The Time of the timeslot you'd like removed.", IsRequired: true},
			},
		},
		{
			Name:        "clear",
			Description: "Clear all your availability rules",
			Action:      availabilityClearAction,
		},
	},
}
